pkg/drivers/powermax.go

```diff
@@ -0,0 +1,258 @@
+	log := logger.GetLogger(ctx)
+	// Check for default secret only
+	// Array specific will be authenticated in csireverseproxy
+	cred := cr.Spec.Driver.AuthSecret
+
+	// Check if driver version is supported by doing a stat on a config file
+	configFilePath := fmt.Sprintf("%s/driverconfig/powermax/%s/upgrade-path.yaml", operatorConfig.ConfigDirectory, cr.Spec.Driver.ConfigVersion)
+	if _, err := os.Stat(configFilePath); os.IsNotExist(err) {
+		log.Errorw("PreCheckPowerMax failed in version check", "Error", err.Error())
+		return fmt.Errorf("%s %s not supported", csmv1.PowerMax, cr.Spec.Driver.ConfigVersion)
+	}
+
+	if cred != "" {
+		found := &corev1.Secret{}
+		err := ct.Get(ctx, types.NamespacedName{Name: cred, Namespace: cr.GetNamespace()}, found)
+		if err != nil {
+			log.Error(err, "Failed query for secret ", cred)
+			if errors.IsNotFound(err) {
+				return fmt.Errorf("failed to find secret %s", cred)
+			}
+		}
+	}
+
+	foundRevProxy := false
+	for _, mod := range cr.Spec.Modules {
+		if mod.Name == csmv1.ReverseProxy {
+			foundRevProxy = true
+			break
+		}
+	}
+	if !foundRevProxy {
+		return fmt.Errorf("failed to find reverseproxy module")
+	}
+
+	log.Debugw("preCheck", "secrets", cred)
+	return nil
+}
+
+// ModifyPowermaxCR -
+func ModifyPowermaxCR(yamlString string, cr csmv1.ContainerStorageModule, fileType string) string {
+	// Parameters to initialise CR values
+	managedArray := ""
+	endpoint := ""
+	clusterPrefix := ""
+	debug := "false"
+	portGroup := ""
+	protocol := ""
+	nodeTemplate := ""
+	modifyHostname := "false"
+	nodeTopology := "false"
+	vsphereEnabled := "false"
+	vspherePG := ""
+	vsphereHostname := ""
+	vsphereHost := ""
+	nodeChap := "false"
+	ctrlHealthMonitor := "false"
+	nodeHealthMonitor := "false"
+	storageCapacity := "false"
+
+	switch fileType {
+	case "Node":
+		for _, env := range cr.Spec.Driver.Common.Envs {
+			if env.Name == "X_CSI_MANAGED_ARRAYS" {
+				managedArray = env.Value
+			}
+			if env.Name == "X_CSI_POWERMAX_ENDPOINT" {
+				endpoint = env.Value
+			}
+			if env.Name == "X_CSI_K8S_CLUSTER_PREFIX" {
+				clusterPrefix = env.Value
+			}
+			if env.Name == "X_CSI_POWERMAX_DEBUG" {
+				debug = env.Value
+			}
+			if env.Name == "X_CSI_POWERMAX_PORTGROUPS" {
+				portGroup = env.Value
+			}
+			if env.Name == "X_CSI_TRANSPORT_PROTOCOL" {
+				protocol = env.Value
+			}
+			if env.Name == "X_CSI_VSPHERE_ENABLED" {
+				vsphereEnabled = env.Value
+			}
+			if env.Name == "X_CSI_VSPHERE_PORTGROUP" {
+				vspherePG = env.Value
+			}
+			if env.Name == "X_CSI_VSPHERE_HOSTNAME" {
+				vsphereHostname = env.Value
+			}
+			if env.Name == "X_CSI_VCENTER_HOST" {
+				vsphereHost = env.Value
+			}
+			if env.Name == "X_CSI_VSPHERE_ENABLED" {
+				vsphereEnabled = env.Value
+			}
+			if env.Name == "X_CSI_IG_MODIFY_HOSTNAME" {
+				modifyHostname = env.Value
+			}
+			if env.Name == "X_CSI_IG_NODENAME_TEMPLATE" {
+				nodeTemplate = env.Value
+			}
+		}
+		for _, env := range cr.Spec.Driver.Node.Envs {
```

Comment:
Dont we want to consider any of env variables specific to controller

Reply:
It is after the Node section @176